File: sdks/go/qnsi/storage/storage.go
// Package storage is the QNSP Storage client - PQC-encrypted object
// storage with SSE-X (server-side encryption with extended PQC).
//
// Wraps `apps/storage-service` (see /proxy/storage in production).
// Requires a billing tier with `sseEnabled: true` (see ClientLimits.SSE).
package storage

import (
	"context"
	"encoding/base64"
	"fmt"
	"net/http"
	"time"

	"github.com/heossihq/qnsi-public/sdks/go/qnsi/internal/qnsicore"
)

// PathPrefix is "/storage/storage/v1" on the edge gateway because the
// edge proxies "/storage" to the storage-service which itself mounts
// routes under "/storage". See edge-gateway proxy table.
const pathPrefix = "/proxy/storage/v1"

type Client struct {
	base *qnsicore.ServiceClient
}

func New(act *qnsicore.Activator, httpClient *http.Client, timeout time.Duration) *Client {
	return &Client{base: qnsicore.NewServiceClient(act, httpClient, pathPrefix, timeout)}
}

// PutObjectRequest is the JSON body for PUT /storage/storage/v1/buckets/:bucket/objects/:key.
type PutObjectRequest struct {
	DataB64      string         `json:"dataB64"`
	ContentType  string         `json:"contentType,omitempty"`
	SSEAlgorithm string         `json:"sseAlgorithm,omitempty"` // e.g. "ml-kem-768"
	Metadata     map[string]any `json:"metadata,omitempty"`
}

func (c *Client) PutObject(ctx context.Context, bucket, key string, req PutObjectRequest, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPut, "/buckets/"+bucket+"/objects/"+key, req, nil, idempotencyKey)
}

// GetObject returns the encrypted object descriptor and a base64-decoded
// payload. The convenience wrapper here only exposes the metadata-and-bytes
// shape used by short objects; for streaming, call Do() directly.
func (c *Client) GetObject(ctx context.Context, bucket, key string) ([]byte, map[string]any, error) {
	resp, err := c.base.Do(ctx, http.MethodGet, "/buckets/"+bucket+"/objects/"+key, nil, nil, "")
	if err != nil {
		return nil, nil, err
	}
	dataB64, _ := resp["dataB64"].(string)
	if dataB64 == "" {
		return nil, resp, fmt.Errorf("qnsp: storage.GetObject: response missing dataB64")
	}
	body, err := base64.StdEncoding.DecodeString(dataB64)
	if err != nil {
		return nil, resp, fmt.Errorf("qnsp: storage.GetObject: dataB64 not valid base64: %w", err)
	}
	return body, resp, nil
}

func (c *Client) DeleteObject(ctx context.Context, bucket, key string) error {
	_, err := c.base.Do(ctx, http.MethodDelete, "/buckets/"+bucket+"/objects/"+key, nil, nil, "")
	return err
}

func (c *Client) ListObjects(ctx context.Context, bucket string, query map[string]string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/buckets/"+bucket+"/objects", nil, query, "")
}

func (c *Client) ListBuckets(ctx context.Context) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/buckets", nil, nil, "")
}

// Do exposes the underlying service client for endpoints not yet wrapped.
func (c *Client) Do(ctx context.Context, method, path string, body any, query map[string]string, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, method, path, body, query, idempotencyKey)
}

File: sdks/rust/qnsi/src/storage.rs
//! QNSI Storage - PQC-encrypted object storage with SSE-X.

use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine;
use reqwest::Method;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::sync::Arc;

use crate::activation::Activation;
use crate::errors::{ApiError, Error};
use crate::service_client::ServiceClient;

// Edge gateway rewrites /proxy/storage -> /storage, so this resolves to the
// backend's /storage/v1/buckets/* routes. (Was /proxy/storage/storage/v1 - a
// double "storage" the backend never served, so every call 404'd. Matches npm.)
const PATH_PREFIX: &str = "/proxy/storage/v1";

#[derive(Clone)]
pub struct Client {
    sc: ServiceClient,
}

impl Client {
    pub fn new(activation: Arc<Activation>, http: reqwest::Client) -> Self {
        Self { sc: ServiceClient::new(activation, http, PATH_PREFIX) }
    }

    pub async fn put_object(&self, bucket: &str, key: &str, req: PutObjectRequest, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc.request(Method::PUT, &format!("/buckets/{}/objects/{}", bucket, key), Some(&req), None, idempotency_key).await
    }

    /// Returns the decoded object bytes alongside the descriptor JSON.
    pub async fn get_object(&self, bucket: &str, key: &str) -> Result<(Vec<u8>, Value), Error> {
        let resp = self.sc.request::<()>(Method::GET, &format!("/buckets/{}/objects/{}", bucket, key), None, None, None).await?;
        let data_b64 = resp
            .get("dataB64")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ApiError {
                status_code: 200,
                code: None,
                message: "storage.get_object: response missing dataB64".into(),
                body: Some(resp.clone()),
            })?;
        let bytes = B64.decode(data_b64).map_err(|_| Error::Api(ApiError {
            status_code: 200,
            code: None,
            message: "storage.get_object: dataB64 not valid base64".into(),
            body: None,
        }))?;
        Ok((bytes, resp))
    }

    pub async fn delete_object(&self, bucket: &str, key: &str) -> Result<(), Error> {
        self.sc.request::<()>(Method::DELETE, &format!("/buckets/{}/objects/{}", bucket, key), None, None, None).await?;
        Ok(())
    }

    pub async fn list_objects(&self, bucket: &str, query: &[(&str, String)]) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, &format!("/buckets/{}/objects", bucket), None, Some(query), None).await
    }

    pub async fn list_buckets(&self) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, "/buckets", None, None, None).await
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PutObjectRequest {
    #[serde(rename = "dataB64")]
    pub data_b64: String,
    #[serde(rename = "contentType", skip_serializing_if = "Option::is_none")]
    pub content_type: Option<String>,
    #[serde(rename = "sseAlgorithm", skip_serializing_if = "Option::is_none")]
    pub sse_algorithm: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Map<String, Value>>,
}
